src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -373,21 +374,43 @@ public IndexMetadata getIndexMetadata()
         return config;
     }
 
+     @Override
+     public boolean shouldSkipInitialization()
+     {
+         // SAI performs partial initialization so it must always execute it; the actual index build is then still skipped
+         // if IndexBuildDecider.instance.onInitialBuild().skipped() is true.
+         return false;
+    }
+
     @Override
     public Callable<?> getInitializationTask()
     {
+        IndexBuildDecider.Decision decision = IndexBuildDecider.instance.onInitialBuild();
         // New storage-attached indexes will be available for queries after on disk index data are built.
         // Memtable data will be indexed via flushing triggered by schema change
         // We only want to validate the index files if we are starting up
-        return () -> startInitialBuild(baseCfs, StorageService.instance.isStarting()).get();
+        return () -> startInitialBuild(baseCfs, StorageService.instance.isStarting(), decision.skipped()).get();
     }
 
-    private Future<?> startInitialBuild(ColumnFamilyStore baseCfs, boolean validate)
+    private Future<?> startInitialBuild(ColumnFamilyStore baseCfs, boolean validate, boolean skipIndexBuild)
     {
+        if (skipIndexBuild)
+        {
+            logger.info("Skipping initialization task for {}.{} after flushing memtable", baseCfs.metadata(), indexContext.getIndexName());
+            // Force another flush to make sure on disk index is generated for memtable data before marking it queryable.
+            // In case of offline scrub, there is no live memtables.
+            if (!baseCfs.getTracker().getView().liveMemtables.isEmpty())
+                baseCfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.USER_FORCED);
```

Comment:
Should probably be INTERNALLY_FORCED or maybe one of INDEX_BUILD_STARTED/INDEX_BUILD_COMPLETED

```suggestion
                baseCfs.forceBlockingFlush(ColumnFamilyStore.FlushReason.INTERNALLY_FORCED);
```

Reply:
I agree. I used `USER_FORCED` because that is what the source in BDP uses. Do you want me to change it?

Reply:
Looks like I made a minor mistake in the cherry pick. It's `INDEX_BUILD_STARTED` a few lines down, fixing now.